### ISSUE TITLE
Incorrect position shift of ligand in simulation

### ISSUE THREAD

**Issue by VincenzoChen** [openmm/openmm#3164]
In some systems I have met such problem that my ligand in systems had a moving which seems not incorrect. Here I take a system as an example.
This system has a relatively closed pocket and the ligand can not pass through the protein to the outside of the pocket. And I also add a distance restraint(white ball in picture one) on ligand mass center and pocket mass center by  CustomCentroidBondForce to make sure the ligand stay in the pocket. However, when I perform the simulation, I find that the ligand run into a position which is outside of the pocket(orange one in picture one) at first frame as if it pass through the protein while in my start structure, the ligand is correctly in the pocket(orange one in picture two). So I wonder how this position shift come into being?
![issue](https://user-images.githubusercontent.com/40376903/124209993-3e425080-db1d-11eb-9434-8ef0fd89f10f.png)
![issue1](https://user-images.githubusercontent.com/40376903/124211382-95492500-db1f-11eb-9c76-5c350901fecc.png)

**Comment by VincenzoChen** [comment]
Another question is whether the CustomCentroidBondForce is dynamics in the simulations. I mean whether the mass center of two group in CustomCentroidBondForce will be claculated every frames. Or they just be calculated before the simulation and never ne calculated again in the simulation?

**Comment by peastman** [comment]
Is it possible this is caused by periodic boundary conditions, and the ligand just seems to jump because it's reporting a different periodic copy?

If that's not it, could you post a working test case so I can reproduce it?  Otherwise I'm just guessing.

**Comment by VincenzoChen** [comment]
[charmm_openmm.tar.gz](https://github.com/openmm/openmm/files/6757761/charmm_openmm.tar.gz)
Here I upload a system that has such problem. The folder includes my scripts to run simulation. I use crystal_ligand.pdb as reference to get the pocket residue(<0.5nm) and use this group and the ligand to define a 0.9nm distance restraint. I have also restrained the CA of protein in the simulation, but the side chain of protein is flexible.

**Comment by peastman** [comment]
I spent some time studying the files.  Were they generated by CHARMM-GUI?  What they're doing is kind of strange.  They don't seem to record the periodic box dimensions anywhere.  Instead they just compute the minimum and maximum coordinate of any atom along each axis and take that to be the box dimensions.  The coordinates also look odd.  For example, there's a single ion whose X coordinate is exactly 0.  But the next lowest X coordinate of any atom is about 17.

I notice that in defining the CustomCentroidBondForce, it specifies

```python
posres.setUsesPeriodicBoundaryConditions(True)
```

That means that distance is calculated while taking periodic boundary conditions into account.  If you displace the ligand by any integer number of box widths, the energy won't change.

Have you investigated whether that's all that's happening?  That the protein and ligand are being reported in different periodic copies?  See https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#periodic for more discussion of this.

**Comment by jchodera** [comment]
> Were they generated by CHARMM-GUI? What they're doing is kind of strange. They don't seem to record the periodic box dimensions anywhere. Instead they just compute the minimum and maximum coordinate of any atom along each axis and take that to be the box dimensions.

Note that we have the same issue with CHARMM-GUI [here](https://github.com/openmm/openmm/issues/3152), and likely need to reach out to the CHARMM-GUI developers to ensure they write out the box vectors for OpenMM into the PDB file.

**Comment by VincenzoChen** [comment]
Yes, I generated my input files by CHARMM-GUI. Actually, the coordinates in the file generated by CHARMM-GUI are all along the axis (see in the figure). So I think the way computing the minimum and maximum coordinate of any atom along each axis has no problem.
![image](https://user-images.githubusercontent.com/40376903/125401576-e8578d80-e3e5-11eb-9a34-2608c4e8dbba.png)

**Comment by peastman** [comment]
> So I think the way computing the minimum and maximum coordinate of any atom along each axis has no problem.

You're going to have a gap of 17 A along the X axis where there's no water, and similar but slightly smaller gaps along the other axes.

> Have you investigated whether that's all that's happening? That the protein and ligand are being reported in different periodic copies?

Let me know what you find for this.

**Comment by jchodera** [comment]
@peastman : Do you want to reach out to the CHARMM-GUI folks about the box vectors issue, or should we hold that until we have a completely specified list of changes needed for full OpenMM compatibility once we resolve #3152?

**Comment by VincenzoChen** [comment]
> Let me know what you find for this.
I have tried another system which has no such gap but the ligand also has a incorrect position shift
![image](https://user-images.githubusercontent.com/40376903/125572294-347f999f-1005-4d25-b1ef-6989f76c820b.png)

**Comment by peastman** [comment]
You need to compare the two locations and see whether they differ by a box vector.
